RoslynPluginGenerator/AnalyzerPluginGenerator.cs

```diff
@@ -61,27 +61,97 @@ public bool Generate(ProcessedArgs args)
-            if (!analyzers.Any())
+            // Check that there are analyzers in the target package from which information can be extracted
+            if (!HasAnalyzers(targetPackage, args.Language))
             {
-                return false;
+                this.logger.LogWarning(UIResources.APG_NoAnalyzersFound, targetPackage.Id);
+
+                if (!args.RecurseDependencies || !dependencyPackages.Any())
+                {
+                    return false;
+                }
+                else
+                {
+                    // Possible sub-case - target package has dependencies that contain analyzers
+                    bool dependencyPackagesHaveAnalyzers = false;
+                    foreach (IPackage dependencyPackage in dependencyPackages)
+                    {
+                        bool dependencyPackageHasAnalyzers = HasAnalyzers(dependencyPackage, args.Language);
+
+                        if (!dependencyPackageHasAnalyzers)
+                        {
+                            this.logger.LogWarning(UIResources.APG_NoAnalyzersFound, dependencyPackage.Id);
+                        }
+
+                        dependencyPackagesHaveAnalyzers = dependencyPackagesHaveAnalyzers || dependencyPackageHasAnalyzers;
+                    }
+
+                    if (!dependencyPackagesHaveAnalyzers)
+                    {
+                        return false;
+                    }
+                }
             }
 
-            IEnumerable<IPackage> licenseAcceptancePackages = this.GetPackagesRequiringLicenseAcceptance(package);
-            if (licenseAcceptancePackages.Any() && ! args.AcceptLicenses)
+            // Check for packages that require the user to accept a license
+            IEnumerable<IPackage> licenseAcceptancePackages = this.GetPackagesRequiringLicenseAcceptance(targetPackage);
+            if (licenseAcceptancePackages.Any() && !args.AcceptLicenses)
             {
-                this.logger.LogError(UIResources.APG_NGPackageRequiresLicenseAcceptance, package.Id, package.Version.ToString());
+                this.logger.LogError(UIResources.APG_NGPackageRequiresLicenseAcceptance, targetPackage.Id, targetPackage.Version.ToString());
                 this.ListPackagesRequiringLicenseAcceptance(licenseAcceptancePackages);
                 return false;
             }
 
+            // Initial run with the user-targeted package and arguments
+            bool result = GeneratePluginForPackage(args, targetPackage);
+            if (!result)
+            {
+                return false;
+            }
+
+            // Dependent package generation changes the arguments
+            if (args.RecurseDependencies)
+            {
+                this.logger.LogWarning(UIResources.APG_RecurseEnabled_SQALENotEnabled);
+
+                foreach (IPackage currentPackage in dependencyPackages)
```

Comment:
You've checked above which packages contain analysers so there's no need to loop through through the entire list again. It would be better to build up a list of packages that contain analysers in the code above. The _dependencyPackagesHaveAnalyzers_ variable wouldn't be needed - you could just check whether the list was empty or not.